### PR TITLE
LUCENE-9515: Detach DWPT from DefaultIndexingChain

### PR DESCRIPTION
**Pull request by s1monw**
This change removes the DWPT dependency from DefaultIndexingChain
and rather passes on the primitives needed for creating the chain.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java [hunk 1]
```diff
@@ -73,21 +78,38 @@
...
-
-  DefaultIndexingChain(DocumentsWriterPerThread docWriter) {
-    this.docWriter = docWriter;
-    this.fieldInfos = docWriter.getFieldInfosBuilder();
-    this.infoStream = docWriter.getIndexWriterConfig().getInfoStream();
+  private final ByteBlockPool.Allocator byteBlockAllocator;
+  private final LiveIndexWriterConfig indexWriterConfig;
+  private final int indexCreatedVersionMajor;
+  private final Consumer<Throwable> abortingExceptionConsumer;
+  private boolean hasHitAbortingException = false;
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:85]: Don't need the `= false` -- it is java's default already.

File: lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java [hunk 2]
```diff
@@ -73,21 +78,38 @@
...
-      storedFieldsConsumer = new SortingStoredFieldsConsumer(docWriter.codec, docWriter.directory, docWriter.getSegmentInfo());
-      termVectorsWriter = new SortingTermVectorsConsumer(docWriter);
+      storedFieldsConsumer = new SortingStoredFieldsConsumer(indexWriterConfig.getCodec(), directory, segmentInfo);
+      termVectorsWriter = new SortingTermVectorsConsumer(intBlockAllocator, byteBlockAllocator, directory, segmentInfo, indexWriterConfig.getCodec());
     }
-    termsHash = new FreqProxTermsWriter(docWriter, bytesUsed, termVectorsWriter);
+    termsHash = new FreqProxTermsWriter(intBlockAllocator, byteBlockAllocator, bytesUsed, termVectorsWriter);
+  }
+
+  private void onAbortingException(Throwable th) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:109]: This was moved from `DWPT`?

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 3]
```diff
@@ -272,7 +273,9 @@ static int getActualMaxDocs() {
    * and a message is printed to infoStream, if set (see {@link
    * IndexWriterConfig#setInfoStream(InfoStream)}).
    */
-  public final static int MAX_TERM_LENGTH = DocumentsWriterPerThread.MAX_TERM_LENGTH_UTF8;
+  // if you increase this, you must fix field cache impl for
+  // getTerms/getTermsIndex requires <= 32768
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:277]: Maybe remove this stale comment :)  FieldCache is long gone!

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java [hunk 4]
```diff
@@ -1533,7 +1533,7 @@ public void testWickedLongTerm() throws IOException {
     Directory dir = newDirectory();
     RandomIndexWriter w = new RandomIndexWriter(random(), dir, new StringSplitAnalyzer());
 
-    char[] chars = new char[DocumentsWriterPerThread.MAX_TERM_LENGTH_UTF8];
+    char[] chars = new char[IndexWriter.MAX_TERM_LENGTH];
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java:1536]: Ooh good catch (`char[]` vs `byte[]`).

**Review by mikemccand — APPROVED** [review]
Looks great!  Thank you for decoupling these two crazy classes @s1monw

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks, Simon!

File: lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java (hunk 2 above)
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:109]: yeah I just tried to not repeat the assignment of `hasHitAbortingException`

**Pull request merged**